vllm/model_executor/layers/quantization/awq_triton.py

```diff
@@ -0,0 +1,296 @@
+
+        b = (b >> shifts) & 0xF
+        zeros = (zeros >> shifts) & 0xF
+        b = (b - zeros) * scales
+        b = b.to(c_ptr.type.element_ty)
+
+        # Accumulate results.
+        accumulator = tl.dot(a, b, accumulator, out_dtype=accumulator_dtype)
+
+        offsets_k += BLOCK_SIZE_K * SPLIT_K
+        a_ptrs += BLOCK_SIZE_K * SPLIT_K
+        b_ptrs += BLOCK_SIZE_K * SPLIT_K * (N // 8)
+
+    c = accumulator.to(c_ptr.type.element_ty)
+    offs_cm = pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+    offs_cn = pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+    c_ptrs = c_ptr + N * offs_cm[:, None] + offs_cn[None, :]
+    c_mask = (offs_cm[:, None] < M) & (offs_cn[None, :] < N)
+    if SPLIT_K == 1:
+        tl.store(c_ptrs, c, mask=c_mask)
+    else:
+        tl.atomic_add(c_ptrs, c, mask=c_mask)
+
+
+# qweights - [K     , M // 8], int32
+# scales   - [K // G, M     ], float16
+# zeros    - [K // G, M // 8], int32
+def awq_dequantize_triton(qweight: torch.Tensor, scales: torch.Tensor,
+                          zeros: torch.Tensor) -> torch.Tensor:
+    K = qweight.shape[0]
+    M = scales.shape[1]
+    awq_group_size = 128
+
+    assert K > 0 and M > 0
+    assert scales.shape[0] == K // awq_group_size and scales.shape[1] == M
+    assert zeros.shape[0] == K // awq_group_size and zeros.shape[1] == M // 8
+
+    # Result tensor:
+    # number of rows = same as input tensor
+    # number of cols = 8 x input tensor num cols
+    result = torch.empty(qweight.shape[0],
+                         qweight.shape[1] * 8,
+                         device=qweight.device,
+                         dtype=scales.dtype)
+
+    block_size_x = 32
+    block_size_y = 32
+
+    Y = qweight.shape[0]  # num rows
+    X = qweight.shape[1]  # num cols
+    group_size = 128
+    grid = lambda META: (
+        triton.cdiv(X, META['BLOCK_SIZE_X']),
+        triton.cdiv(Y, META['BLOCK_SIZE_Y']),
+    )
+    awq_dequantize_kernel[grid](qweight,
+                                scales,
+                                zeros,
+                                group_size,
+                                result,
+                                X,
+                                Y,
+                                BLOCK_SIZE_X=block_size_x,
+                                BLOCK_SIZE_Y=block_size_y)
+
+    return result
+
+
+# input   - [M, K]
+# qweight - [K, N // 8]
+# qzeros  - [K // G, N // 8]
+# scales  - [K // G, N]
+# split_k_iters - parallelism along K-dimension, int, power of 2.
+def awq_gemm_triton(input: torch.Tensor, qweight: torch.Tensor,
+                    scales: torch.Tensor, qzeros: torch.Tensor,
+                    split_k_iters: int) -> torch.Tensor:
+    M, K = input.shape
+    N = qweight.shape[1] * 8
+    awq_group_size = 128
+
+    assert N > 0 and K > 0 and M > 0
+    assert qweight.shape[0] == K and qweight.shape[1] == N // 8
+    assert qzeros.shape[0] == K // awq_group_size and qzeros.shape[1] == N // 8
```

Comment:
I would like to see an explicit assert calculating the group size from the packed qzeros and asserting if it isn't 128, if you will only support group size 128 for this kernel. It needs to be clear that this is the only supported config. It might even be best to do this check with `is_hip()` in the awq.py config

Reply:
Assert added